### PR TITLE
[WIP] Add support for platform-defined standard directories

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

File: src/cargo/util/config.rs [hunk 1]
```diff
@@ -916,7 +964,7 @@ fn walk_tree<F>(pwd: &Path, mut walk: F) -> CargoResult<()>
     // Once we're done, also be sure to walk the home directory even if it's not
     // in our history to be sure we pick up that standard location for
     // information.
-    let home = homedir(pwd).ok_or_else(|| {
+    let home = cargo_home_dir(pwd).ok_or_else(|| {
```
Comment (matklad) [inline src/cargo/util/config.rs]: This should use `config_dir` rather than `homedir` I think?

File: src/cargo/util/config.rs [hunk 2]
```diff
@@ -896,7 +944,7 @@ impl fmt::Display for Definition {
     }
 }
 
-pub fn homedir(cwd: &Path) -> Option<PathBuf> {
+fn cargo_home_dir(cwd: &Path) -> Option<PathBuf> {
     ::home::cargo_home_with_cwd(cwd).ok()
```
Comment (matklad) [inline src/cargo/util/config.rs]: Interesting! This `cargo_home_with_cwd` has a lot of code for multirust -> rustup transition. I wonder if we should remove most of it?

File: src/cargo/util/config.rs [hunk 3]
```diff
@@ -83,10 +90,9 @@ impl Config {
         });
 
         Config {
-            home_path: Filesystem::new(homedir),
+            dirs: dirs,
```
Comment (matklad) [inline src/cargo/util/config.rs]: might be just `dirs` instead of `dirs: dirs`

File: src/cargo/util/config.rs (hunk 2 above)
Comment (soc) [reply src/cargo/util/config.rs]: Not sure what's the time-line on this, but probably yes. The whole concept of a "cargo root" has to go anyway, it's just not meaningful. There _could_ be a cargo root (if `CARGO_ROOT` is set), but there is no guarantee.

What needs to be done is to replace everything that uses the cargo root to derive directories with calls to more specific functions (e. g. ask directly for the config dir, or the cache dir, etc.).

**Comment by soc** [comment]
@matklad thanks for having a look! Yes, other code definitely needs to be adapted, too.

I will also add a `cargo dirs` command that returns the directories to help tools that work on the command line. So that stuff like `$CARGO_ROOT/bin` can be replaced with `$(cargo dirs executable)` or something.

**Pull request closed without merging**